### PR TITLE
fix: SQL Lab sorting of non-numbers

### PR DESCRIPTION
**Pull request by etr2460**
### SUMMARY
https://github.com/apache/superset/pull/17573 fixed the sorting of floating point numbers, but broke sorting strings that start with numbers. The most common of these are dses like `2022-01-01`, and `parseFloat("2022-01-01") === 2022`, so sorting got all janky.

I've fixed this by ensuring that we only try to parse numbers if they pass a regex that matches JS's format of numbers.

Note that this solution isn't 100% accurate, and will still cause undefined behavior when sorting a list containing both numeric and non numeric strings. I don't have a good answer for fixing this, but perhaps in the future we'll need to refactor this further. I'd imagine a per column parameter that defines the sorting function (convert to number and then sort, do string sort, sort in hexadecimal because all the strings start with `0x` and contain only `0-f`, something else) but that was very much out of scope of this bug fix.

### TESTING INSTRUCTIONS
New unit test (and ensure the unit test added in #17573 still works)

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

to: @lyndsiWilliams @eschutho @graceguo-supercat @ktmud

### CODE REVIEW

File: superset-frontend/src/components/FilterableTable/FilterableTable.tsx [hunk 1]
```diff
@@ -322,16 +322,26 @@ export default class FilterableTable extends PureComponent<
...
-    const floatValue = parseFloat(value);
-    return Number.isNaN(floatValue) ? value : floatValue;
+  // Parse any numbers from strings so they'll sort correctly
+  parseNumberFromString = (value: string | number | null) => {
+    if (typeof value === 'string') {
+      // This regex handles all possible number formats in javascript, including ints, floats,
+      // exponential notation, NaN, and Infinity.
+      // See https://stackoverflow.com/a/30987109 for more details
+      const onlyNumberRegEx =
+        /^(NaN|-?((\d*\.\d+|\d+)([Ee][+-]?\d+)?|Infinity))$/;
```
Comment (ktmud) [inline superset-frontend/src/components/FilterableTable/FilterableTable.tsx:332]: Maybe move this to a global constant or create a helper function?

**Review by ktmud — APPROVED** [review]
LGTM with a small suggestion.

Did you check whether the performance is acceptable when table has the maximum number of rows? I'd imagine regex match for every row to be quite expensive. In regards to the RegEx used, it's probably more important what the database returns other than what JavaScript accepts. For example, some database may return `Infinity` as `Inf`.

I think in long run we would want to apply different sorting function based on column types. We should pass down the column type prop to this component and apply number parsing only to numeric columns. We may also want to migrate this table to [react-table](https://react-table.tanstack.com/docs/examples/virtualized-rows) as well. It has a [sortType](https://react-table.tanstack.com/docs/api/useSortBy#column-options) configuration we can use for free.

**Comment by etr2460** [comment]
>Did you check whether the performance is acceptable when table has the maximum number of rows? I'd imagine regex match for every row to be quite expensive. In regards to the RegEx used, it's probably more important what the database returns other than what JavaScript accepts. For example, some database may return Infinity as Inf.

@ktmud I did not check yet, although aren't regexes pretty fast usually? I suppose i could do a faster check first before the more expensive regex, but my thought was that the `parseFloat` was already taking a while (that we run on each iteration) so maybe it didn't matter. Regardless I'll test a large result set and report back.

As for matching what databases may return, I suppose i could try to handle everything every database returns, but i figured having something consistent with a single language would be worthwhile. If this was breaking for you with a specific database, then you could always write `if(column_name = Inf, 'Infinity', column_name) as column_name` in your query, while if we don't have this handling, it could never sort properly (or only would if you did something like `if(column_name = Inf, 99999999999999999999, column_name) as column_name`

So in summary, i'll test perf and if that looks good, will merge (with the moving of the regex to a global constant, sorry for not doing that in the first place)

**Review by eschutho — APPROVED** [review]
Thanks for catching/fixing this issue @etr2460!

**Comment by etr2460** [comment]
verified that 10k rows still sort pretty much instantly, both for numeric and non numeric columns
